### ISSUE TITLE
Extend routing.xml to represent that some point features are good, not bad

### ISSUE THREAD

**Issue by david-gpu** [osmandapp/OsmAnd#21320]
### Describe the idea

`routing.xml` should have the ability to represent that routes that contain certain point features are more desireable, not less. The existing "obstacle" section can only be used to penalize certain points, but it doesn't allow for the opposite.

Let me illustrate with an example: while a car driver may consider that a speed hump, a chicane or some other traffic calming feature is bad because it slows them down, it is typically good for a cycling route, *precisely because* it makes drivers slow down.

### Expected behaviour

A backwards-compatible extension to `routing.xml` would be to add an optional attribute to the `select` element, such that it reverses the sign of the `value` attribute. This would allow a cycling profile to contain something like this:

```
        <!-- Penalty used by the routing algorithm but NOT displayed to the user -->
        <point attribute="obstacle">
            <select value="10" t="traffic_calming" sign="-1"/> <!-- Crude blanket boost to ways that include traffic calming measures -->
        </point>
```

### Alternatives you've considered

A terrible alternative would be assigning a small penalty to every single point feature, but that would mean that if a way contains a high density of points then it would be severely penalized against another way that is just as long but contains fewer points.

**Comment by david-gpu** [comment]
This task is not complete because the existing functionality clearly does not meet the requirements of this proposed feature.

As explained in the section titled *"Alternatives that you've considered"*, we can't simply penalize every point in a way as a way of guiding the routing algorithm towards routes that include certain point features such as traffic calming elements. It just doesn't work, because two ways that cover the same distance may be penalized very differently simply based on the number of points in those ways.

> Obviously negative penalties are not possible cause it will break routing concept as creating minimal cost (cost can't be negative infinitely) route

The routing cost can be trivially capped to zero with something like `cost = max(cost, 0)`. The proposal is made to be simple to implement and backwards compatible.

Want to call it "nice to have" or "will not fix"? Fine. But please don't mark it a s complete because this functionality is not implemented today and can't be simulated with the existing functionality.

**Comment by david-gpu** [comment]
> You can do already obstacles with penalty=0, I see no issue with that.

Let's use a representative example to see the limitations of the current code.

Let's say you are going out for a bicycle ride and the streets look as follows:

![image](https://github.com/user-attachments/assets/1fead124-16fd-4a1f-a41a-cc5dce94f701)

Both streets have the same length, but the street on the left has no features and the street on the right has several traffic calming measures. Which one would you rather ride on your bike with your 8yo daughter? How would you write `routing.xml` so that in the following scenario the route on the right will be selected? A penalty of zero is not going to help in any way.

Today, there is nothing in the `routing.xml` schema that will allow us to represent that the street on the right is preferable, because `routing.xml` doesn't understand the obvious fact that some point features are *better* than nothing.

**Comment by david-gpu** [comment]
Incidentally, implementing this would also address part of the problem alluded to in [Issue #21050](https://github.com/osmandapp/OsmAnd/issues/21050), since an intersection with traffic lights is safer than one without.

**Comment by vshcherb** [comment]
Yes but that's not possible from routing algorithm perspective at all. You can't simply predict non-existing features while exploring other routes. 
The only possibility is to make alternative routes and than select. 

This one is also very-very tricky https://github.com/osmandapp/OsmAnd/issues/21050, but at least I see some possibility in future. But for this request I simply don't understand how it's possible, it's one of the basic limitation for Dijkstra which is the slowest and base algorithm for others faster. 

You can't have negative penalties on the route

**Comment by david-gpu** [comment]
I can only understand your comment if the current routing algorithm is so naive that each little segment in OSM is directly mapped to one edge in the routing graph, and the same is done for points. Is that the case?

Until now I have assumed that since it is not possible to make a (meaningful) routing decision until an intersection is found, the routing algorithm would aggregate the cost of all sequential segments in a `way` until a fork/intersection is found. This would incidentally make the graph a ton smaller and routing computations a ton faster.

If cost was aggregated in the space between intersections, it would be trivial to cap it to zero as suggested earlier. It doesn't require changing the A* algorithm.

I spent some time looking at the source code, but gave up after I couldn't find any source-level documentation.

**Comment by vshcherb** [comment]
>  simple that each little segment in OSM is directly mapped to one edge in the routing graph, and the same is done for points. Is that the case

With some assumptions - yes, and then A* is running with assigned penalties on the edges of course

**Comment by vshcherb** [comment]
If you can explain how your example works with Dijkstra animation somehow it would be enough cause I simply don't understand from theoritical point of view how it could be fixed. https://en.m.wikipedia.org/wiki/File:Dijkstra_Animation.gif

**Comment by david-gpu** [comment]
OK. Let's start with the basics, before any of the changes I'm proposing.

Here is a real-world street map taken from OSM.

![image](https://github.com/user-attachments/assets/5e68e639-2986-4e24-8d19-7243a4d8d08d)

Now let's draw a graph on top of it. The starting node is painted green, the destination node is painted orange, all other graph nodes are painted in pink, and the edges are painted in yellow.

![image](https://github.com/user-attachments/assets/b104d95f-d68e-4e09-8a62-918df8f6c3a2)

For simplicity, let's assume that the penalty for all of the nodes (points) is zero. The only routing cost comes from the **9** yellow edges.

The naive Dijkstra algorithm would only traverse one edge at a time and compute the cost at each node, then continue from the node that is closest to the starting node if it has any neighbors that haven't been visited yet. Do we agree so far?

But I don't think that is how the existing code works, at least in [the C++ version](https://github.com/osmandapp/OsmAnd-core/blob/ada80c0ec1b22bcb078898bda9932eec1ff429c7/src/Routing/RoutePlanner.cpp#L784). Maybe I'm mistaken -- please correct me if I'm wrong. I believe that the current code is more efficient and traverses multiple OSM edges in sequence [as long as they belong to the same segment](https://github.com/osmandapp/OsmAnd-core/blob/ada80c0ec1b22bcb078898bda9932eec1ff429c7/src/Routing/RoutePlanner.cpp#L858), then stops at intersections (i.e. points with more than two ways).

This makes a lot of sense -- it reduces the number of times at which you have to evaluate which node to evaluate next by simply staying on the same road until an intersection is found. Once you enter a road, either you find your destination or you keep going until the next intersection where you can go left/right/forward.

That is, you could say that the routing algorithm in practice looks at the graph like this, with **7** nodes rather than the original **9**, as the penalties of the two traffic calming nodes on Elm Ridge Drive on the south are simply aggregated to the cost of the edges they are connected to.

![image](https://github.com/user-attachments/assets/5c6470c6-e48b-4dea-94a4-a10480108304)

Isn't that how the existing code works?

**Comment by vshcherb** [comment]
> The naive Dijkstra algorithm would only traverse one edge at a time and compute the cost at each node, then continue from the node that is closest to the starting node if it has any neighbors that haven't been visited yet. Do we agree so far?

Yes exactly how it works. You wanted to put a negative cost to edge - that's not possible. If you say I want to have "reduced cost" for edge cause edge cost consists = Penalty for points + Distance / Speed. So it's not negative, than it requires very special graph construction somehow

**Comment by david-gpu** [comment]
>> The naive Dijkstra algorithm would only traverse one edge at a time and compute the cost at each node, then continue from the node that is closest to the starting node if it has any neighbors that haven't been visited yet. Do we agree so far?
>  Yes exactly how it works.

So the Java code no longer follows ways until it finds an intersection like the old C++ code did? Okay, that's interesting. It doesn't make a big difference to what we are proposing here, but it will make it less effective.

> You wanted to put a negative cost to edge - that's not possible. If you say I want to have "reduced cost" for edge cause edge cost consists = Penalty for points + Distance / Speed. So it's not negative, than it requires very special graph construction somehow

Sorry I haven't been clear. I do not want to assign negative costs to edges. I want to assign negative costs to `point` features, which would indeed allow for `penalty_for_points` to take a negative value, but because the total cost would be then clamped to zero, the total cost would be always non-negative. In terms of the expression you mentioned, it would look like this:

```
cost = max(0, penalty_for_points + distance/speed)
```

It never becomes negative, but it allows `penalty_for_points` to have negative values.

Bringing this back to the real-world example I provided earlier, we would have something like the following:

**Comment by vshcherb** [comment]
Let me introduce slightly more details. So we can finally reach full picture.

In any routing algorithm (theoritical) you can't put any cost to the node! You need to put cost only to the edges. Though the trick is to do transformation with graph so routing algorithm could be applied. Of course negative cost still not possible on edges.

If you have situation ```A -(cost X)-> B (cost Y) -(cost Z)-> C ```. You introduce points (Vertices) in the middle of each segment and connect with Edges (all possible directions) : ```B1 (middle AB) --B (cost = X / 2 + Y + Z / 2)--> B2 (middle BC)```. So it's kind of OSM points become Edges and OSM segments become Vertices.

You are right with description 
```cost = max(0, penalty_for_points + penalty_for_distance) = max(0, -3 + 2) = max(0,-1) = 0```, so technically it's possible in that situation. So I reopen the issue.

However your solution is still very cumbersome:
1. Requires thorough refactoring ( less of an issue) - as that feature was never prepared
2. Your negative penalty very-very dependent on the length - cost of the OSM segment. So it could be tiny and same features will have different impact on the rought in different places

**Comment by vshcherb** [comment]
So for me it looks very strange still, probably it will have same effect as preprocessing roads and put higher priority if there are any specific points on them. For example introducing artificial tag "road_with_traffic_calming"
